### PR TITLE
Faction infighting update

### PR DESCRIPTION
**Pull request by Coolthulhu**
Inter-faction relations

Factions are now loaded from a json. Each faction has (in addition to name and id) a base faction and an attitude map.

Attitude map maps factions to one of (currently 3) faction attitudes: BY_MOOD, NEUTRAL and FRIENDLY. BY_MOOD is the default attitude: hostile if angry, neutral otherwise. Neutral and friendly don't check for anything.

If a target faction is not in our attitude map, its base faction is checked instead. Then base faction of base faction and so on until a sentinel faction is reached. Sentinel faction provides default behavior: BY_MOOD for all factions except vermin (neutral for those).

Base faction is also used for finalization: each faction inherits base faction's attitude map.
For example, "cult" is a subfaction of "zombie", which is a subfaction of "" (the sentinel faction).

Example of usage: police bot (factions cop_bot, subfaction of bot) and zombie cop (faction cop_zombie, subfaction of zombie) see each other.
Faction cop_bot is always friendly to cop_zombie. cop_zombie doesn't override any behavior when it comes to bots, so defaults to BY_MOOD.
Police bot is friendly to police zed. Police zed is hostile to the bot.

Factions that are neutral/friendly to each other are totally skipped when making plans. This should speed up game as most things will want to ignore vermin.
Own faction isn't specifically checked for and is only set as friendly by default (can be overridden in json). As an example, I set jabberwocks to be hostile towards their own kind.

When loading factions from jsons, a graph is generated and then loaded starting from the sentinel faction. Any faction without an explicit parent is treated as a subfaction of the sentinel faction.
Bad jsons (containing dependency cycles) should not lock up the game, but produce a debug message and break the cycle by re-parenting cyclical factions to sentinel.

I disabled infighting for vermin for now. They slowed down the game quite a bit with all this needless planning. They still look for player, but only for player.

### CODE REVIEW

**Comment by Rivet-the-Zombie** [comment]
Cool stuff!

**Comment by KA101** [comment]
Objection: bot operators were mistaken for hostiles as pert of the cataclysm, so bots shouldn't be friendly to folks from "their" faction.

**Comment by Rivet-the-Zombie** [comment]
How does this square with eyebots being friendly towards police characters?

**Comment by Coolthulhu** [comment]
Disabled copbot->copzed alliance.
Police bots generated through means other than eyebots (is it even possible?) will attack police survivors, so it makes sense for them to attack police zeds too.
When eyebots start targetting zeds, it will need a rework.

**Comment by KA101** [comment]
Eyebots specifically validate the ID: my thought was that being pure surveillance models, they *weren't* reprogrammed to be more aggressive.

(They aren't Friendly! per se, just I rewrote their special attack.)

File: src/monstergenerator.cpp [hunk 1]
```diff
@@ -53,6 +54,69 @@ void MonsterGenerator::finalize_mtypes()
...
+    nullfaction->base_faction = nullptr;
+    // Create a tree of faction dependence
+    std::multimap< const monfaction*, monfaction* > child_map;
+    std::set< monfaction* > unloaded; // To check if cycles exist
+    std::queue< monfaction* > queue;
+    for( auto &pair : faction_map ) {
+        auto faction = &pair.second;
+        unloaded.insert( faction );
+        // Point parent to children
+        child_map.emplace( faction->base_faction, faction );
```
Comment (KA101) [inline src/monstergenerator.cpp]: C::B can't find emplace in std::multimap.  Sorry, Coolthulhu.

Comment (Coolthulhu) [inline src/monstergenerator.cpp]: That's strange

std::multimap::emplace is part of the core std lib and it worked just fine with gcc 4.9.1.

Switched for manual pair creation and then inserting that. It should do the same, it just looks less clean.

Comment (KA101) [inline src/monstergenerator.cpp]: Ah, one of those things.  Problem is that C::B officially defaults to 4.7.1.  One of those minimize-hurdles things.

**Pull request merged**